ClinGen/clincoded
# 1669 - when a gdm is created, a classification is provisioned, approved, published or unpublished, send related data to Data Exchange.

Steps to test:
1. Log in and associate with an affiliation that can publish GDM (e.g. ClinGen Curator Test Affiliation 1).
2. Create a new Gene-Disease Record (GDM).
3a. Check the debug console, shoud see a log message "Post tracking data succeeded:..." with sent json message.
3b. If testing locally and have set up kafka server locally (see instructions below), should see the creation json message appears in the consumer terminal.
3c. If there is error, should see log message "Post tracking data failed:..." and other "Error sending ..." message.
    And an error message started with "** track-data: Error sending ..." is logged to apache error log.
4. Create some evidences (with different users from same affiliation if possible).
5. Go to "Classification Matrix" page, save the classification.
6. Provisionally approve the classification.
7. Messages and sent json message should be logged as in step 3.
8. Approve the classification, see logged messages and json message as in step 3.
9. Do publish and unpublish classification and check json message as in step 3.

Set up kafka server to view publish and track_data messages locally:
1. Set up kafka server locally
https://kafka.apache.org/quickstart
Follow step 1 through step 5
After step 1, change the default port from 9092 to 9093
From installed directory, edit config/server.properties
        listeners=PLAINTEXT://:9092 => listeners=PLAINTEXT://:9093
Then continue to other steps
- To purge old data, delete and create the topic again.
bin/kafka-topics.sh --bootstrap-server localhost:9093 --delete --topic test
bin/kafka-topics.sh --create --bootstrap-server localhost:9093 --replication-factor 1 --partitions 1 --topic test
2. In /src/clincoded/__init__.py, uncommented following
"# config.include('.messaging')"
3. Install confluent_kafka
https://github.com/confluentinc/confluent-kafka-python/issues/405
> pip install confluent_kafka
Or
>pip install confluent_kafka==0.11.4
4. In /src/clincoded/messaging/__init__.py, under both publish and track_data codes,
def publish(request):
def track_data(request):
Changed default kafka_conf settings to
    kafka_conf = {'bootstrap.servers': 'localhost:9093',
            'log_level': 0}
Only left 'log_level' and commented/removed the rest

More technical details in https://asjira.stanford.edu/jira/browse/CLINGEN-3
